File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.py
def multiply(a, b):
    n = len(a)
    temp = [[0] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            for k in range(n):
                temp[i][j] += a[i][k] * b[k][j]
                temp[i][j] %= 1000

    return temp


def solve(arr, b):
    if b == 1:
        return arr

    half = solve(arr, b // 2)
    result = multiply(half, half)

    if b % 2 == 1:
        result = multiply(result, arr)
    return result

n, b = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]

result = solve(arr, b)

for row in result:
    print(" ".join(map(lambda x: str(x % 1000), row)))

File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    final static int MOD = 1000;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        long b = Long.parseLong(st.nextToken());

        int[][] arr = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] result = solve(arr, b);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sb.append(result[i][j] % MOD).append(" ");
            }
            sb.append("\n");
        }

        System.out.print(sb);
    }

    private static int[][] solve(int[][] arr, long b) {
        if (b == 1L){
            return arr;
        }

        int[][] half = solve(arr, b / 2);
        int[][] result = multiply(half, half);

        if (b % 2 == 1L) {
            result = multiply(result, arr);
        }

        return result;
    }

    private static int[][] multiply(int[][] a, int[][] b) {
        int n = a.length;
        int[][] temp = new int[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    temp[i][j] += a[i][k] * b[k][j];
                    temp[i][j] %= MOD;
                }
            }
        }

        return temp;
    }
}